wcfsetup/install/files/lib/system/package/PackageInstallationDispatcher.class.php

```diff
@@ -350,11 +355,46 @@ public function install($node) {
 				WHERE		processNo = ?";
 			$statement = WCF::getDB()->prepareStatement($sql);
 			$statement->execute([$this->queue->processNo]);
+			
+			$this->logInstallationStep([], 'finished cleanup');
 		}
 		
 		return $step;
 	}
 	
+	/**
+	 * Logs an installation step.
+	 * 
+	 * @param	array		$node	data of the executed node
+	 * @param	string		$log	optional additional log text
+	 */
+	protected function logInstallationStep(array $node = [], $log = '') {
+		$time = TIME_NOW;
+		$logEntry = "[" . TIME_NOW . "]\n";
+		if (!empty($node)) {
+			$logEntry .= 'sequenceNo: ' . $node['sequenceNo'] . "\n";
+			$logEntry .= 'nodeType: ' . $node['nodeType'] . "\n";
+			$logEntry .= "nodeData:\n";
+			
+			$nodeData = unserialize($node['nodeData']);
+			foreach ($nodeData as $index => $value) {
+				$logEntry .= "\t" . $index . ': ' . (!is_object($value) && !is_array($value) ? $value : JSON::encode($value)) . "\n";
```

Comment:
Are you positive that JSON encoding objects does not lose information (such as private fields)?

Reply:
Objects are actually irrelevant here because only arrays are present here as a "non-printable" data type. I just added the object check for completeness.